packages/@o3r/schematics/src/utility/dependencies.ts

```diff
@@ -32,12 +54,200 @@ export function getExternalDependenciesVersionRange<T extends string>(packageNam
+export function isRangeGreater(rangeA: string, rangeB: string) {
+  if (rangeA === 'latest' && rangeB !== 'latest') {
+    return true;
+  }
+  if (rangeB === 'latest') {
+    return false;
+  }
+  const minVersionA = minVersion(rangeA)!;
+  const minVersionB = minVersion(rangeB)!;
+  const isBPreleaseOfVersionA = minVersionB.prerelease?.length > 0
+    && subset(`${minVersionB.major}.${minVersionB.minor}.${minVersionB.patch}`, rangeA);
+  return subset(rangeB, rangeA) || isBPreleaseOfVersionA || (gt(minVersionA, minVersionB) && !subset(rangeA, rangeB));
+}
+
+/**
+ * Compute the version range for a package based on the less restrictive version declared in the package.json.
+ * If the versions declared in the package.json do not intersect, take the highest version.
+ * @param packageName
+ * @param packageJsonContent
+ * @param isTildeEnforced
+ */
+export function getDependencyMaximumVersionRange(packageName: string, packageJsonContent: PackageJson, isTildeEnforced?: boolean): string {
+  const rangeSortedByHighestMinimum = [
+    packageJsonContent.peerDependencies?.[packageName],
+    packageJsonContent.dependencies?.[packageName],
+    packageJsonContent.devDependencies?.[packageName]
+  ]
+    .map((packageVersion) => isTildeEnforced && packageVersion ? enforceTildeRange(packageVersion) : packageVersion)
+    .filter((packageVersion): packageVersion is string => {
+      return valid(packageVersion) !== null || validRange(packageVersion) !== null;
+    })
+    .sort((rangeA, rangeB) =>
+      isRangeGreater(rangeB, rangeA) ? 1 : (isRangeGreater(rangeA, rangeB) ? -1 : 0)
+    );
+  return rangeSortedByHighestMinimum[0];
+}
+
+/**
+ * Find the range for this package based on the generatorDependency object of the package.json
+ * If there are no specified generator dependency, look for the range with the highest minimum or the widest range
+ * @param packageName
+ * @param packageJsonPath
+ * @param isTildeEnforced
+ */
+export function getVersionToInstallFromPackageJson(packageName: string, packageJsonPath: string, isTildeEnforced?: boolean): string | undefined {
+  const packageJsonContent = (
+    JSON.parse(fs.readFileSync(packageJsonPath, { encoding: 'utf8' }))
+  ) as PackageJson & { generatorDependencies?: Record<string, string> };
+  return packageJsonContent.generatorDependencies?.[packageName]
+    ?? getDependencyMaximumVersionRange(packageName, packageJsonContent, isTildeEnforced);
+}
+
+/**
+ * Check if a dependency has already been installed for the requested range
+ * @param depName
+ * @param packageJson
+ * @param root0
+ * @param root0.range
+ * @param root0.types
+ */
+export function isDependencyAlreadyInstalled(depName: string, packageJson: PackageJson, {
+  range,
+  types
+}: Required<DependencyInManifest>) {
+  const packageJsonDepVersions = types?.map((type) => (packageJson[type] || {})[depName]);
+  return packageJsonDepVersions.some((version) => !!version && (satisfies(version, range) || subset(version, range)));
```

Comment:
is it working properly with prereleases?
```suggestion
  return packageJsonDepVersions.some((version) => !!version && (satisfies(version, range, {includePrerelease: true}) || subset(version, range, {includePrerelease: true})));
```
I think this is needed if we want to check for example:
```typescript
satisfies('12.1.0-rc.0', '^12.0.0'); // <-- false
satisfies('12.1.0-rc.0', '^12.0.0', {includePrerelease: true}); // <-- true
```